WordPress/openverse
# Add color mode to ui store

## Fixes



Fixes #4308 by @zackkrida 

## Description



Core functionality:

- Updated the `dark_mode_ui_toggle` feature flag description to reflect that it now causes the frontend to respect system preferences for color themes. Now when the color mode is "system", no body css class will be set.
- Added media queries in the tailwind.css to adjust the color scheme based on the system's prefers-color-scheme setting.
- Introduced a new `colorMode` state in the ui store which stores the color mode in a cookie.
- Added an action to set the color mode, which the toggle in #4309 will use.
- Updated the `useDarkMode` composable to derive the color mode from the UI store when the `dark_mode_ui_toggle` is enabled, and to override it with the `force_dark_mode` flag when enabled.
- Updated and expanded unit tests for the `useDarkMode` composable and the ui store to cover the new `colorMode` logic.

I also made some small improvements:

- Refactored the UI store's initial state setup to use a new defaultUiState object, which can be used to initialize tests as well.
- Updated type definitions in the feature flag store for stricter type checking (FlagName instead of string for various methods).
- Converted the ui store test to TypeScript and fixed some types.

Here is a video outlining the feature flag behavior:




## Testing Instructions


- Review the code changes and test cases to make sure they are sufficient
- Enable the preference to force dark mode; observe that it is still respected
- Apply this diff to add a rudimentary theme changer to the footer:
```diff
diff --git a/frontend/src/components/VFooter/VFooter.vue b/frontend/src/components/VFooter/VFooter.vue
index ca326f8f4..24d699757 100644
--- a/frontend/src/components/VFooter/VFooter.vue
+++ b/frontend/src/components/VFooter/VFooter.vue
@@ -14,6 +14,7 @@ import VBrand from "~/components/VBrand/VBrand.vue"
 import VLanguageSelect from "~/components/VLanguageSelect/VLanguageSelect.vue"
 import VPageLinks from "~/components/VHeader/VPageLinks.vue"
 import VWordPressLink from "~/components/VHeader/VWordPressLink.vue"
+import { useFeatureFlagStore } from "#imports"
 
 /**
  * The footer is the section displayed at the bottom of a page. It can contain
@@ -72,11 +73,24 @@ export default defineComponent({
       "--link-col-height": Math.ceil(Object.keys(allPages).length / 2),
     }))
 
+    const featureFlagStore = useFeatureFlagStore()
+    const showToggle = computed(() =>
+      featureFlagStore.isOn("dark_mode_ui_toggle")
+    )
+    const colorMode = ref(uiStore.colorMode)
+    const handleColorModeChange = () => {
+      uiStore.setColorMode(colorMode.value)
+    }
+
     return {
       isContentMode,
       allPages,
       currentPage,
 
+      showToggle,
+      colorMode,
+      handleColorModeChange,
+
       footerEl,
       variantNames,
       linkColumnHeight,
@@ -110,6 +124,17 @@ export default defineComponent({
 
     
     <div class="locale-and-wp flex flex-col justify-between">
+      <select
+        class="bg-tx"
+        v-if="showToggle"
+        v-model="colorMode"
+        @change="handleColorModeChange"
+        name="colorMode"
+      >
+        <option value="light">Light</option>
+        <option value="dark">Dark</option>
+        <option value="system">System</option>
+      </select>
       <VLanguageSelect
         v-bind="languageProps"
         class="language max-w-full border-secondary"
```
- Enable the `dark_mode_ui_toggle` flag on the preferences page.
- Try each setting in the footer theme toggle and make sure:
    - Choice is persisted with the correct CSS class on "body" and the correct value in the Ui cookie (check devtools for this)
    - The "system" setting respects if you toggle dark mode at your OS or browser level 

## Checklist


      PRs) or the media properties generator (`ov just catalog/generate-docs media-props`
      for the catalog or `ov just api/generate-docs` for the API) where applicable.

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
